### PR TITLE
Add Debug APIs that return Thread and Query Resource Usage

### PR DESCRIPTION
**Pull request by vrajat**
Query cancellation in OOM conditions depends on tracking resource usage by query execution threads. Resource tracking is implemented in `PerQueryCPUMemAccountantFactory.java`. A couple of debug APIs have been added to broker and server to access the internal data structures that track resource usage. These debug APIs will help to understand the behavior of this feature in production systems. 

The following two APIs have been added in brokers and servers

### debug/queries/resourceUsage

Example response
```
[
  {
    "cpuTimeNs": 0,
    "queryId": "Broker_192.168.0.107_8000_174733410000000095_O",
    "allocatedBytes": 3239944
  },
  {
    "cpuTimeNs": 0,
    "queryId": "Broker_192.168.0.107_8000_174733410000000094_O",
    "allocatedBytes": 3239944
  },
  {
    "cpuTimeNs": 0,
    "queryId": "Broker_192.168.0.107_8000_174733410000000093_O",
    "allocatedBytes": 3239944
  },
  {
    "cpuTimeNs": 0,
    "queryId": "Broker_192.168.0.107_8000_174733410000000082_O",
    "allocatedBytes": 4520488
  },
  {
    "cpuTimeNs": 0,
    "queryId": "Broker_192.168.0.107_8000_174733410000000092_O",
    "allocatedBytes": 2599672
  },
  {
    "cpuTimeNs": 0,
    "queryId": "Broker_192.168.0.107_8000_174733410000000087_O",
    "allocatedBytes": 3880216
  },
  {
    "cpuTimeNs": 0,
    "queryId": "Broker_192.168.0.107_8000_174733410000000088_O",
    "allocatedBytes": 2599672
  },
  {
    "cpuTimeNs": 0,
    "queryId": "Broker_192.168.0.107_8000_174733410000000096_O",
    "allocatedBytes": 1959400
  }
]
```

### /debug/threads/resourceUsage

Example o/p

```
[
  {
    "taskId": -1,
    "queryId": "174733410000001465",
    "cputimeMS": 0,
    "allocatedBytes": 0
  },
  {
    "taskId": -1,
    "queryId": "174733410000001466",
    "cputimeMS": 0,
    "allocatedBytes": 0
  },
  {
    "taskId": -1,
    "queryId": "174733410000001467",
    "cputimeMS": 0,
    "allocatedBytes": 0
  },
  {
    "taskId": -1,
    "queryId": "Broker_192.168.0.107_8000_174733410000001466_O",
    "cputimeMS": 0,
    "allocatedBytes": 0
  },
  {
    "taskId": 0,
    "queryId": "Broker_192.168.0.107_8000_174733410000001466_O",
    "cputimeMS": 0,
    "allocatedBytes": 3239680
  },
]
```
tags: observability

### CODE REVIEW

**Comment by siddharthteotia** [comment]
Will these debug APIs add any performance overhead ?

**Comment by jasperjiaguo** [comment]
A general question -- what is the rational behind adding this?

**Comment by vrajat** [comment]
@siddharthteotia The APIs will not add any perf overhead as the changes are not in path of query execution. 
@jasperjiaguo The APIs will help to introspect internal state to understand OOM protection in production clusters. 
I've found the APIs to be helpful for debugging as well. I am also working on adding OOM protection for multi-stage operators in #13598 and the APIs are a fast way to get internal state while running a workload. 

@jasperjiaguo how do you debug OOM protection and resource usage in production systems right now?

**Comment by jasperjiaguo** [comment]
Yes I also do not see any overhead. Will finish reviewing tmr. IIUC this is trying to expose the per query and per thread snapshot right?

**Comment by vrajat** [comment]
> IIUC this is trying to expose the per query and per thread snapshot right?

Yes. Only on a specific server or a broker. There is no API right now to aggregate across servers or brokers.

**Review by jasperjiaguo — APPROVED** [review]
nit but mostly lgtm

File: pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotBrokerDebug.java [hunk 1]
```diff
@@ -268,4 +273,24 @@ public String getServerRoutingStats() {
...
+  @Authorize(targetType = TargetType.CLUSTER, action = Actions.Cluster.DEBUG_RESOURCE_USAGE)
+  @ApiOperation(value = "Get resource usage of threads")
+  public Collection<? extends ThreadResourceTracker> getThreadResourceUsage() {
+    ThreadResourceUsageAccountant threadAccountant = Tracing.getThreadAccountant();
+    return threadAccountant.getThreadResources();
+  }
+
+  @GET
+  @Path("debug/queries/resourceUsage")
+  @Produces(MediaType.APPLICATION_JSON)
```
[the remark this answers is not in the document]
Comment (vrajat) [reply pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotBrokerDebug.java:289]: Missed it. Added.

File: pinot-tools/src/main/java/org/apache/pinot/tools/utils/PinotConfigUtils.java [hunk 2]
```diff
@@ -215,6 +216,27 @@ public static Map<String, Object> generateMinionConf(String clusterName, String
...
+
+    configOverrides.put(
+        CommonConstants.PINOT_QUERY_SCHEDULER_PREFIX + "." + CommonConstants.Accounting.CONFIG_OF_FACTORY_NAME,
+        PerQueryCPUMemAccountantFactory.class.getCanonicalName());
+    configOverrides.put(CommonConstants.PINOT_QUERY_SCHEDULER_PREFIX + "."
+        + CommonConstants.Accounting.CONFIG_OF_ENABLE_THREAD_MEMORY_SAMPLING, true);
+    configOverrides.put(CommonConstants.PINOT_QUERY_SCHEDULER_PREFIX + "."
+        + CommonConstants.Accounting.CONFIG_OF_ENABLE_THREAD_CPU_SAMPLING, false);
+    configOverrides.put(CommonConstants.PINOT_QUERY_SCHEDULER_PREFIX + "."
+        + CommonConstants.Accounting.CONFIG_OF_OOM_PROTECTION_KILLING_QUERY, true);
```
[the remark this answers is not in the document]
Comment (vrajat) [reply pinot-tools/src/main/java/org/apache/pinot/tools/utils/PinotConfigUtils.java:236]: This config is used in the two Quickstarts I added to try out the debug apis. I havent engineered the workload to simulate OOM protection. That is why I havent specified any thresholds.

**Comment by vrajat** [comment]
The test failure is not related to these changes.

**Pull request merged**